package/src/ssr/NextSSRApolloClient.tsx

```diff
@@ -0,0 +1,169 @@
+    uninitialized: true,
+  };
+
+  constructor(options: ApolloClientOptions<TCacheShape>) {
+    super({
+      ...options,
+      // TODO: the memoization in `documentTransform` can be removed once
+      // https://github.com/apollographql/apollo-client/pull/10968 is merged.
+      documentTransform: new DocumentTransform((document) => {
+        const stringified = print(document);
+        if (seenDocuments.has(stringified)) {
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          return seenDocuments.get(stringified)!;
+        }
+        seenDocuments.set(stringified, document);
+        return document;
+      }),
+    });
+
+    this.registerWindowHook();
+  }
+  private resolveFakeQueries = new Map<
+    string,
+    [(result: FetchResult) => void, (reason: any) => void]
+  >();
+
+  private identifyUniqueQuery(options: {
+    query: DocumentNode;
+    variables?: unknown;
+  }) {
+    const transformedDocument = this.documentTransform.transformDocument(
+      options.query
+    );
+
+    // Calling `transformDocument` will add __typename but won't remove client
+    // directives, so we need to get the `serverQuery`.
+    const { serverQuery } =
+      this["queryManager"].getDocumentInfo(transformedDocument);
+
+    const canonicalVariables = canonicalStringify(options.variables);
+
+    const cacheKey = [serverQuery, canonicalVariables].toString();
+
+    return { query: serverQuery, cacheKey, varJson: canonicalVariables };
+  }
+
+  private registerWindowHook() {
+    if (typeof window !== "undefined") {
+      if (Array.isArray(window[ApolloBackgroundQueryTransport] || [])) {
+        registerLateInitializingQueue(
+          ApolloBackgroundQueryTransport,
+          (options) => {
+            const { query, varJson, cacheKey } =
+              this.identifyUniqueQuery(options);
+
+            const byVariables =
+              this["queryManager"].inFlightLinkObservables.get(query) ||
+              new Map();
+
+            this["queryManager"].inFlightLinkObservables.set(
+              query,
+              byVariables
+            );
+
+            if (!byVariables.has(varJson)) {
+              const promise = new Promise<FetchResult>((resolve, reject) => {
+                this.resolveFakeQueries.set(cacheKey, [resolve, reject]);
+              });
+
+              byVariables.set(
+                varJson,
+                new Observable<FetchResult>((observer) => {
+                  promise
+                    .then((result) => {
+                      observer.next(result);
+                      observer.complete();
+                    })
+                    .catch((err) => {
+                      observer.error(err);
+                    });
+                })
+              );
+
+              const cleanupCancelFn = () =>
+                this["queryManager"].fetchCancelFns.delete(cacheKey);
+
+              const [_, reject] = this.resolveFakeQueries.get(cacheKey) ?? [];
```

Comment:
This line should probably better go into the `fetchCancelFn` definition a few lines down - otherwise `reject` would never be undefined, and also execute if the query had resolved successfully.

Reply:
Fixed in [c49d481](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/c49d481ff5c9bdd474c4b61ba65b610162ffe067).